### DISCUSSION TITLE
How to properly deal with 3-level nesting

### DISCUSSION THREAD

**Discussion by KnowYourselves** [prisma/prisma#6642] [Q&A] [answered]
Hi!

I have a case where I have an instance A that has a one to many relationship with B, and B has a one to many relationship with C.

My question is, Is there a way to directly connect A to C?, like for example, changing all C instances related to a specific A instance.

I can't seem to find a nice way to do it with the API.

Thanks!

**Comment by 2color** [comment, marked as answer]
> My question is, Is there a way to directly connect A to C?, like for example, changing all C instances related to a specific A instance.

In this scenario, the relation between A to C is implicit through table B which means you need to create rows/records in the B table to link a record in A to C.

You can always directly define a relation between A to C. However, whether that's a good idea for you requires a closer look at the schema and problem you are trying to solve.

Perhaps you could share the Prisma schema and explain the problem you're trying to solve.

For example,

Given the following schema:

```prisma
model User {
  id        Int        @id @default(autoincrement())
  name      String     @unique
  nicknames Nickname[]
}

model Nickname {
  id     Int    @id @default(autoincrement())
  name   String @unique
  user   User   @relation(fields: [userId], references: [id])
  userId Int
  memes  Meme[]
}

model Meme {
  id         Int      @id @default(autoincrement())
  name       String   @unique
  nickname   Nickname @relation(fields: [nicknameId], references: [id])
  nicknameId Int
}
```


You could also add a direct relation between `User` and `Meme`:

```prisma
model User {
  id        Int        @id @default(autoincrement())
  name      String     @unique
  nicknames Nickname[]
  memes      Meme[]
}

model Nickname {
  id     Int    @id @default(autoincrement())
  name   String @unique
  user   User   @relation(fields: [userId], references: [id])
  userId Int
  memes  Meme[]
}

model Meme {
  id         Int      @id @default(autoincrement())
  name       String   @unique
  nickname   Nickname @relation(fields: [nicknameId], references: [id])
  nicknameId Int
  user       User     @relation(fields: [userId], references: [id])
  userId     Int
}
```

But that's not really desired, because now you have two ways to link a `User` to a `Meme`

**Reply by KnowYourselves** [reply]
Thanks for the response!

Yeah the schema you show is pretty much the situation, it's nothing really complicated. I thought about making a direct A to C connection but it seems redundant like you say.

My problem is more with the Prisma API, for example, I know that with an A instance I can use findUnique to find all instances of B, but I can't then also add to the query one more level to also get all C instances.

I ended up getting all B instances through A and then using `Promise.all` to get all C instances. I guess my question if this is the expected way to do this kind of query or if there's something I missing from the API.

Thanks!

**Reply by 2color** [reply]
I thought it's worth pointing out that you can both create and query relations of relations in a single query.

Given the schema above, you can do the following:

```ts
  // Create a user, related nicknames, and related memes
  const user1 = await prisma.user.create({
    data: {
      name: 'Alice',
      nicknames: {
        create: [
          {
            name: 'lili',
          },
          {
            name: 'ali',
            memes: {
              create: [
                {
                  name: 'wow',
                },
                { name: 'haha' },
              ],
            },
          },
        ],
      },
    },
  })
  console.log(JSON.stringify(user1, null, '\t'))
  
  // query 3-levels (2 relations between three models)
  const user = await prisma.user.findUnique({
    where: {
      name: 'Alice',
    },
    include: {
      nicknames: {
        include: {
          memes: true,
        },
      },
    },
  })
```


I hope that helps you reduce the number of queries/calls.

**Reply by KnowYourselves** [reply]
Well, color me surprised, I could swear I tried **exactly that** and it gave me an error, but now it just worked.

That's miles better than using `Promise.all`, thanks a ton!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
